### ISSUE TITLE
Kernel 6.5.4 fails to boot on Dell XPS 15 9560

### ISSUE THREAD

**Issue by danielwe** [pop-os/pop#3152]
**Distribution (run `cat /etc/os-release`):**

```
NAME="Pop!_OS"
VERSION="22.04 LTS"
ID=pop
ID_LIKE="ubuntu debian"
PRETTY_NAME="Pop!_OS 22.04 LTS"
VERSION_ID="22.04"
HOME_URL="https://pop.system76.com"
SUPPORT_URL="https://support.system76.com"
BUG_REPORT_URL="https://github.com/pop-os/pop/issues"
PRIVACY_POLICY_URL="https://system76.com/privacy"
VERSION_CODENAME=jammy
UBUNTU_CODENAME=jammy
LOGO=distributor-logo-pop-os
```

**Related Application and/or Package Version (run `apt policy $PACKAGE NAME`):**

```
linux-image-generic:
  Installed: 6.5.4.76060504.202309191142~1695998943~22.04~070916d
  Candidate: 6.5.4.76060504.202309191142~1695998943~22.04~070916d
  Version table:
 *** 6.5.4.76060504.202309191142~1695998943~22.04~070916d 1001
       1001 http://apt.pop-os.org/release jammy/main amd64 Packages
        100 /var/lib/dpkg/status
     5.15.0.86.83 500
        500 http://apt.pop-os.org/ubuntu jammy-security/main amd64 Packages
        500 http://apt.pop-os.org/ubuntu jammy-updates/main amd64 Packages
     5.15.0.25.27 500
        500 http://apt.pop-os.org/ubuntu jammy/main amd64 Packages
```

**Issue/Bug Description:**

My Dell XPS 15 9560 fails to boot after the most recent kernel update. I just get dropped to the initramfs prompt. (Luckily the `oldkern` image still works.)

**Steps to reproduce (if you know):**

* Update to the most recent kernel
* Reboot

**Expected behavior:**

* Successful boot

**Other Notes:**

* Clearly this is likely to be an upstream issue
* There's a report on reddit of a Dell XPS 13 user losing all sound devices after updating the kernel to the same version: https://www.reddit.com/r/pop_os/comments/171lgvz/no_sound_devices_after_upgrading_to_kernel_654/. No idea if that could be related, but it's the same line of laptops.

**Comment by alvins711** [comment]
I'm having the same issue on a Dell Precision 5520. Looking at the dmesg output it's says nvme0 controller is down and can't mount the root partition /dev/nvme0n1p3. 

Tried adding boot option "nvme_core.default_ps_max_latency_us=0 pcie_aspm=off" lets it get past the initramfs but then console loops on reporting "systemd-journald failed to write entry input/output error".

Rolled back to kernel v6.4.6 and it's fine.

**Comment by leviport** [comment]
Can you share the models of drives in these machines?

**Comment by danielwe** [comment]
```
THNSN5512GPUK NVMe TOSHIBA 512GB (5KDA4103)
```

**Comment by alvins711** [comment]
Kingston ‎SKC3000S 1024G Nvme M.2

**Comment by Sky4Viper** [comment]
same here on my Dell Precision 5520

 can only boot by going to old kernel and manually setting it as preferred kernel, 

sudo kernelstub -v -k /boot/vmlinuz-6.4.6-76060406-generic -i /boot/initrd.img-6.4.6-76060406-generic

unfortunately updates keep trying to go to a new kernel every few days, braking the boot, really annoying.

**Comment by leviport** [comment]
Setting oldkern as the default in systemd-boot will work better, since regenerating initramfs won't just undo the change you're making there with kernelstub. Hold spacebar while booting to summon the systemd-boot menu, move the cursor to the oldkern entry, and hit the `d` key to set a default.

**Comment by hoshsadiq** [comment]
Also got the same issue, also a Dell XPS 15 9560
Disk: `KXG50ZNV512G NVMe TOSHIBA 512GB`
Happy to post anything else you need. For now, I've marked the old kernel as the default.

**Comment by alvins711** [comment]
I tried running the latest version of the installer and it couldn't detect the drive to install on. Also checked using GParted and Disks and same result, drive is not being displayed. Ran lsblk from the terminal and it shows the nvme device but does not display the existing partitions.

Out of curiosity, I ran the previous version of the installer and it picks up the drive and partitions just fine.

![Screenshot from 2023-10-12 22-4
![Screenshot from 2023-10-12 22-55-05](https://github.com/pop-os/pop/assets/12243660/cb0bf260-1208-4a52-b637-912d6faf3563)
6-44](https://github.com/pop-os/pop/assets/12243660/b61e1bb9-0abc-42b8-9247-510f6753aa77)

**Comment by danielwe** [comment]
I think this is the relevant upstream issue: https://bugzilla.kernel.org/show_bug.cgi?id=217802

Looks like the fix will land in kernel 6.5.6.

**Comment by danielwe** [comment]
The issue is probably best summarized in this update from the kernel regression tracker: https://lore.kernel.org/all/169557219938.3206394.2779757887621800924@leemhuis.info/

In the meantime, other distros are shipping patched kernels that revert the offending commit:

* Fedora: https://gitlab.com/cki-project/kernel-ark/-/commit/80c615ec2edb4aadded21fe924e2caa172d59577
* OpenSUSE: https://github.com/openSUSE/kernel-source/commit/1b02b1528a26f4e9b577e215c114d8c5e773ee10
* NixOS: https://github.com/NixOS/nixpkgs/pull/255824

**Comment by leviport** [comment]
Since I don't have the hardware to recreate this, could someone use the [Mainline tool](https://github.com/bkw777/mainline) on Pop to check that the 6.5.6 Ubuntu kernel fixes the problem? There is another fix in 6.5.6 that we've been looking for as well, so it'd be nice to know if 6.5.6 fixes this too, or if it landed later in Ubuntu's mainline kernel.

**Comment by Sky4Viper** [comment]
what would be the proper url for adding minline ppa?
sudo add-apt-repository ppa:cappelikan/ppa
times out for me

**Comment by Sky4Viper** [comment]
what would be the proper url for adding minline ppa?
sudo add-apt-repository ppa:cappelikan/ppa
times out for me

**Comment by alvins711** [comment]
I'm getting a 404 error on mainline ppa url - https://kernel.ubuntu.com/~kernel-ppa/mainline/, not loading any kernel apart from the local ones.

**Comment by danielwe** [comment]
I couldn't use mainline in pop because I have the Nvidia driver and VirtualBox installed, but I did some experiments on my Arch partition. In summary:

* 6.5.3 does not work, as expected: boot crashes when trying to mount filesystems
* 6.5.6 works!
* 6.5.7 also works

These Arch kernels only have a few patches applied on top of upstream, none of which are related to this issue, and most of which are identical between versions. You can see for yourself here (look at the commits between the "Arch Linux kernel" by heftig at the top and "Linux" by gregkh just a few commits down):

* 6.5.3: https://github.com/archlinux/linux/commits/v6.5.3-arch1
* 6.5.6: https://github.com/archlinux/linux/commits/v6.5.6-arch2

In conclusion: **Yes, the issue is fixed in 6.5.6.**

(In contrast, the Arch releases of 6.5.4 and 6.5.5 had reverted the offending commit, so they too work on my laptop. You can see the revert right below the top commit here: https://github.com/archlinux/linux/commits/v6.5.4-arch2. Neither 6.5.3, 6.5.6, or 6.5.7 included the revert, so the results above must be explained by upstream changes.)

**Comment by danielwe** [comment]
If someone can point me to a Pop installation ISO with kernel <=6.4.10 I can set up a separate Pop partition so I can test the Ubuntu mainline kernels. (The current ISO from the Pop website uses kernel 6.5.4, so I can't install from it because of this same bug.)

**Comment by leviport** [comment]
We're currently testing a 6.5.6 PR: https://github.com/pop-os/linux/pull/288 
After I've done a little smoke testing, I'll post commands others can use to install the staging branch, if anyone wants to follow along and help test.

@danielwe you can change the version number in the download URL to get older versions of the ISO. The current version is 35, and I believe 34 had 6.4.6

**Comment by danielwe** [comment]
Confirming that an up-to-date Pop!_OS 22.04 boots successfully with the Ubuntu mainline 6.5.6 kernel.

**Comment by leviport** [comment]
I've not completed testing yet, but from a quick first look, it's looking good. If anyone wants to follow along and help test, you can add the staging branch with this command:
```
sudo apt-manage add popdev:linux-6.5.6
```
and then install updates normally.

Once that PR is merged, the staging branch should be removed with this command:
```
apt-manage remove popdev-linux-6-5-6
```
(it's formatted slightly differently when you remove it, but it should tab-complete)

**Comment by hoshsadiq** [comment]
@leviport I was going to test this, so I ran the upgrade command, then thought about the issue mentioned in https://github.com/pop-os/kernelstub/issues/76, then saw that same issue. I'm a little worried about rebooting to test now. Is there a way to make sure I have a good kernel to fallback on in case it doesn't work?

**Comment by danielwe** [comment]
@hoshsadiq what's the output of `sudo kernelstub -v -c`? (This is a dry-run command, it won't change your system.)

**Comment by hoshsadiq** [comment]
```#$ sudo kernelstub -v -c
kernelstub.Config    : INFO     Looking for configuration...
kernelstub           : INFO     System information:

    OS:..................Pop!_OS 22.04
    Root partition:....../dev/dm-1
    Root FS UUID:........cd5b6c1a-48a1-47f1-9450-79bc47227fcc
    ESP Path:............/boot/efi
    ESP Partition:......./dev/nvme0n1p1
    ESP Partition #:.....1
    NVRAM entry #:.......-1
    Boot Variable #:.....0000
    Kernel Boot Options:.splash quiet acpi_rev_override=5 loglevel=0 systemd.show_status=false pci=nommconf nouveau.modeset=0
    Kernel Image Path:.../boot/vmlinuz-6.5.6-76060506-generic
    Initrd Image Path:.../boot/initrd.img-6.5.6-76060506-generic
    Force-overwrite:.....False

kernelstub.Installer : INFO     Copying Kernel into ESP
kernelstub.Installer : INFO     Simulate copying: /boot/vmlinuz-6.5.6-76060506-generic => /boot/efi/EFI/Pop_OS-cd5b6c1a-48a1-47f1-9450-79bc47227fcc/vmlinuz.efi
kernelstub.Installer : INFO     Copying initrd.img into ESP
kernelstub.Installer : INFO     Simulate copying: /boot/initrd.img-6.5.6-76060506-generic => /boot/efi/EFI/Pop_OS-cd5b6c1a-48a1-47f1-9450-79bc47227fcc/initrd.img
kernelstub.Installer : INFO     Setting up loader.conf configuration
kernelstub.Installer : INFO     Simulate creation of entry...
kernelstub.Installer : INFO     Loader entry: /boot/efi/loader/entries/Pop_OS-current
title Pop!_OS
linux /EFI/Pop_OS-cd5b6c1a-48a1-47f1-9450-79bc47227fcc/vmlinuz.efi
initrd /EFI/Pop_OS-cd5b6c1a-48a1-47f1-9450-79bc47227fcc/initrd.img
options root=UUID=cd5b6c1a-48a1-47f1-9450-79bc47227fcc ro splash quiet acpi_rev_override=5 loglevel=0 systemd.show_status=false pci=nommconf nouveau.modeset=0

kernelstub.Installer : INFO     Backing up old kernel
kernelstub.Installer : INFO     Simulate copying: /boot/vmlinuz-6.5.4-76060504-generic => /boot/efi/EFI/Pop_OS-cd5b6c1a-48a1-47f1-9450-79bc47227fcc/vmlinuz-previous.efi
kernelstub.Installer : INFO     Simulate copying: /boot/initrd.img-6.5.4-76060504-generic => /boot/efi/EFI/Pop_OS-cd5b6c1a-48a1-47f1-9450-79bc47227fcc/initrd.img-previous
kernelstub.Installer : INFO     Making entry file for Pop!_OS
kernelstub.Installer : INFO     Simulate copying: /proc/cmdline => /boot/efi/EFI/Pop_OS-cd5b6c1a-48a1-47f1-9450-79bc47227fcc
```

**Comment by danielwe** [comment]
Yeah, that means the broken kernel is now in your `oldkern`. Luckily, I can report that 6.5.6 from the staging branch works flawlessly on my system, but it's probably a good idea to fix this before you reboot. Here are the steps I would take (note that this is 100 % unofficial support, I have no affiliation with System76, I'm just a user like you):

* Verify that kernel 6.4.6 is still installed: `apt-cache policy linux-image-6.4.6-76060406-generic linux-headers-6.4.6-76060406-generic` (the output should indicate that these packages are installed)
* `sudo apt remove linux-image-6.5.4* linux-headers-6.5.4*`
* `sudo kernelstub -v -c` (dry-run, same as above)
  * verify that the lines under "Backing up old kernel" are copying things that have version 6.4.6
* `sudo kernelstub -v` (actual run)

Now you should have the 6.5.6 staging kernel in `current` and 6.4.6 in `oldkern` and rebooting is hopefully safe.

**Comment by hoshsadiq** [comment]
Aah, perfect! Thank you.

**Comment by alvins711** [comment]
@leviport after a quick test, I can confirm that 6.5.6 now recognizes the partitions on my machine.

**Comment by leviport** [comment]
6.5.6 has just been released. Thank you everyone who reported the bug and helped test the fix!
